src/app/loadout-builder/utils.ts

```diff
@@ -74,3 +78,109 @@ export function getPower(items: DimItem[] | ProcessItem[]) {
+ */
+function getMaxEnergyFromUpgradeSpendTier(
+  defs: D2ManifestDefinitions,
+  tier: UpgradeSpendTier,
+  item: DimItem
+) {
+  if (!item.energy) {
+    return 0;
+  }
+
+  const isExotic = Boolean(item.equippingLabel);
+  // Used for figuring out what upgrade is not allowed
+  // if undefined there is no boundary
+  let boundaryHash: number | 'none' = 'none';
+
+  switch (tier) {
+    case UpgradeSpendTier.Nothing:
+      // special case, value is 0 so items energy wins out or swap is always disallowed
+      return 0;
+    case UpgradeSpendTier.LegendaryShards:
+      boundaryHash = UpgradeMaterialHashes.enhancementPrism;
+      break;
+    case UpgradeSpendTier.EnhancementPrisms:
+      boundaryHash = UpgradeMaterialHashes.ascendantShard;
+      break;
+    case UpgradeSpendTier.AscendantShardsNotExotic: {
+      if (!isExotic) {
+        break;
+      }
+      boundaryHash = UpgradeMaterialHashes.ascendantShard;
+      break;
+    }
+    case UpgradeSpendTier.AscendantShards:
+      break;
+  }
+
+  const availableEnergyUpgrades = energyUpgrade(
+    defs,
+    item,
+    item.energy.energyType,
+    item.energy.energyCapacity,
+    item.energy.energyType,
+    10
+  );
+
+  // Just get the max possible energy capacity for the item.
+  if (boundaryHash === 'none') {
+    return Math.max(
+      item.energy.energyCapacity,
+      ...availableEnergyUpgrades.map(
+        (upgradeHash) => defs.InventoryItem.get(upgradeHash).plug!.energyCapacity!.capacityValue
+      )
+    );
+  }
+
+  let previousUpgrade: DestinyInventoryItemDefinition | undefined;
+  for (const upgrade of availableEnergyUpgrades) {
+    const upgradeItem = defs.InventoryItem.get(upgrade);
+    const upgradeMaterials = defs.MaterialRequirementSet.get(
+      upgradeItem.plug!.insertionMaterialRequirementHash
+    );
+
+    for (const material of upgradeMaterials.materials) {
+      if (material.itemHash === boundaryHash) {
+        // in the case of no previous upgrade the item must already be on the limit
+        return previousUpgrade
+          ? previousUpgrade.plug!.energyCapacity!.capacityValue
+          : item.energy.energyCapacity;
+      }
+    }
+    previousUpgrade = upgradeItem;
+  }
+
+  // should never happen but lets be safe
+  return 0;
+}
+
+/** Gets the max energy allowed from the passed in UpgradeSpendTier */
+export function upgradeSpendTierToMaxEnergy(
+  defs: D2ManifestDefinitions,
+  tier: UpgradeSpendTier,
+  item: DimItem
+) {
+  if (!item.energy) {
+    return 0;
+  }
+
+  return Math.max(item.energy.energyCapacity, getMaxEnergyFromUpgradeSpendTier(defs, tier, item));
+}
+
+/** Figures out whether you can swap energies in the allowed spend tier. */
+export function canSwapEnergyFromUpgradeSpendTier(
+  defs: D2ManifestDefinitions,
+  tier: UpgradeSpendTier,
+  item: DimItem
+) {
+  if (!item.energy) {
+    return false;
+  }
+  return item.energy.energyCapacity <= getMaxEnergyFromUpgradeSpendTier(defs, tier, item);
```

Comment:
For the swap you may want to pass a different energy type in to the `energyUpgrade` function

Reply:
It would make a difference if we were showing the actual cost for changing energy but it doesn't effect the results if all we care about is boundary elements like I do at the moment. In saying that SHOULD bungie change it so changing elements is cheaper it would future proof the solution.